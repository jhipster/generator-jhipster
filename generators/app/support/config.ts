import { camelCase, kebabCase, startCase, upperFirst } from 'lodash-es';
import { NODE_VERSION } from '../../generator-constants.js';
import { applicationTypes, authenticationTypes, databaseTypes, testFrameworkTypes } from '../../../lib/jhipster/index.js';
import { getHipster, mutateData, pickFields, upperFirstCamelCase } from '../../base/support/index.js';
import { getDBTypeFromDBValue } from '../../server/support/index.js';
import detectLanguage from '../../languages/support/detect-language.js';
import { loadConfig, loadDerivedConfig } from '../../../lib/internal/index.js';
import serverCommand from '../../server/command.js';
import { packageJson } from '../../../lib/index.js';
import type { ApplicationType } from '../../../lib/types/application/application.js';

const { GATLING, CUCUMBER, CYPRESS } = testFrameworkTypes;
const { GATEWAY, MONOLITH } = applicationTypes;
const { JWT, OAUTH2, SESSION } = authenticationTypes;
const { NO: NO_DATABASE } = databaseTypes;

/**
 * Load common options to be stored.
 * @deprecated
 */
export function loadStoredAppOptions(this: any, { options = this.options, jhipsterConfig = this.jhipsterConfig, log = this.log } = {}) {
  // Parse options only once.
  if (this.sharedData.getControl().optionsParsed) return;
  this.sharedData.getControl().optionsParsed = true;

  if (options.db) {
    const databaseType = getDBTypeFromDBValue(options.db);
    if (databaseType) {
      jhipsterConfig.databaseType = databaseType;
    } else if (!jhipsterConfig.databaseType) {
      throw new Error(`Could not detect databaseType for database ${options.db}`);
    }
    jhipsterConfig.devDatabaseType = options.db;
    jhipsterConfig.prodDatabaseType = options.db;
  }
  if (options.testFrameworks) {
    jhipsterConfig.testFrameworks = [...new Set([...(jhipsterConfig.testFrameworks || []), ...options.testFrameworks])];
  }
  if (options.language) {
    // workaround double options parsing, remove once generator supports skipping parse options
    const languages = options.language.flat();
    if (languages.length === 1 && languages[0] === 'false') {
      jhipsterConfig.enableTranslation = false;
    } else {
      jhipsterConfig.languages = [...(jhipsterConfig.languages || []), ...languages];
    }
  }
  if (options.nativeLanguage) {
    if (typeof options.nativeLanguage === 'string') {
      jhipsterConfig.nativeLanguage = options.nativeLanguage;
      if (!jhipsterConfig.languages) {
        jhipsterConfig.languages = [options.nativeLanguage];
      }
    } else if (options.nativeLanguage === true) {
      jhipsterConfig.nativeLanguage = detectLanguage();
    }
  }

  if (jhipsterConfig.clientPackageManager) {
    const usingNpm = jhipsterConfig.clientPackageManager === 'npm';
    if (!usingNpm) {
      log?.warn(`Using unsupported package manager: ${jhipsterConfig.clientPackageManager}. Install will not be executed.`);
      options.skipInstall = true;
    }
  }
}

/**
 * Load app configs into application.
 * all variables should be set to dest,
 * all variables should be referred from config,
 * @param {any} config - config to load config from
 * @param {any} dest - destination context to use default is context
 */
export const loadAppConfig = ({
  config,
  application,
  useVersionPlaceholders,
}: {
  config: any;
  application: ApplicationType;
  useVersionPlaceholders?: boolean;
}) => {
  loadConfig(serverCommand.configs, { config, application });

  mutateData(
    application,
    {
      nodeVersion: useVersionPlaceholders ? 'NODE_VERSION' : NODE_VERSION,
      jhipsterVersion: useVersionPlaceholders ? 'JHIPSTER_VERSION' : undefined,
    },
    pickFields(config, [
      'jhipsterVersion',
      'baseName',
      'reactive',
      'jhiPrefix',
      'skipFakeData',
      'entitySuffix',
      'dtoSuffix',
      'skipCheckLengthOfIdentifier',
      'microfrontend',
      'microfrontends',
      'skipServer',
      'skipCommitHook',
      'skipClient',
      'prettierJava',
      'pages',
      'skipJhipsterDependencies',
      'withAdminUi',
      'gatewayServerPort',
      'capitalizedBaseName',
      'dasherizedBaseName',
      'humanizedBaseName',
      'projectDescription',
      'authenticationType',
      'rememberMeKey',
      'jwtSecretKey',
      'fakerSeed',
      'skipUserManagement',
      'blueprints',
      'testFrameworks',
      'syncUserWithIdp',
    ]),
    {
      jhipsterVersion: packageJson.version,
      blueprints: [],
      testFrameworks: [],
    },
  );
};

/**
 * @param {Object} dest - destination context to use default is context
 */
export const loadDerivedAppConfig = ({ application }: { application: any }) => {
  loadDerivedConfig(serverCommand.configs, { application });

  mutateData(application, {
    jhiPrefixCapitalized: ({ jhiPrefix }) => upperFirst(jhiPrefix),
    jhiPrefixDashed: ({ jhiPrefix }) => kebabCase(jhiPrefix),

    camelizedBaseName: ({ baseName }) => camelCase(baseName),
    hipster: ({ baseName }) => getHipster(baseName),
    capitalizedBaseName: ({ baseName }) => upperFirst(baseName),
    dasherizedBaseName: ({ baseName }) => kebabCase(baseName),
    lowercaseBaseName: ({ baseName }) => baseName?.toLowerCase(),
    upperFirstCamelCaseBaseName: ({ baseName }) => upperFirstCamelCase(baseName),
    humanizedBaseName: ({ baseName }) => (baseName.toLowerCase() === 'jhipster' ? 'JHipster' : startCase(baseName)),

    gatlingTests: ({ testFrameworks }) => testFrameworks?.includes(GATLING),
    cucumberTests: ({ testFrameworks }) => testFrameworks?.includes(CUCUMBER),
    cypressTests: ({ testFrameworks }) => testFrameworks?.includes(CYPRESS),

    projectDescription: ({ projectDescription, baseName }) => projectDescription ?? `Description for ${baseName}`,
    endpointPrefix: ({ applicationType, lowercaseBaseName }) => (applicationType === 'microservice' ? `services/${lowercaseBaseName}` : ''),
  } as any);

  if (application.microfrontends && application.microfrontends.length > 0) {
    application.microfrontends.forEach(microfrontend => {
      const { baseName } = microfrontend;
      microfrontend.lowercaseBaseName = baseName.toLowerCase();
      microfrontend.capitalizedBaseName = upperFirst(baseName);
      microfrontend.endpointPrefix = `services/${microfrontend.lowercaseBaseName}`;
    });
  } else if (application.microfrontend) {
    application.microfrontends = [];
  }
  application.microfrontend =
    application.microfrontend ||
    (application.applicationTypeMicroservice && !application.skipClient) ||
    (application.applicationTypeGateway && application.microfrontends && application.microfrontends.length > 0);

  if (application.microfrontend && application.applicationTypeMicroservice && !application.gatewayServerPort) {
    application.gatewayServerPort = 8080;
  }

  application.authenticationTypeSession = application.authenticationType === SESSION;
  application.authenticationTypeJwt = application.authenticationType === JWT;
  application.authenticationTypeOauth2 = application.authenticationType === OAUTH2;

  application.generateAuthenticationApi = application.applicationType === MONOLITH || application.applicationType === GATEWAY;
  const authenticationApiWithUserManagement = application.authenticationType !== OAUTH2 && application.generateAuthenticationApi;
  application.generateUserManagement =
    !application.skipUserManagement && application.databaseType !== NO_DATABASE && authenticationApiWithUserManagement;
  application.generateInMemoryUserCredentials = !application.generateUserManagement && authenticationApiWithUserManagement;
};
