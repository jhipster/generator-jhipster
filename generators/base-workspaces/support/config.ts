import { mutateData, pickFields } from '../../../lib/utils/index.ts';

import { monitoringTypes, serviceDiscoveryTypes } from '../../../lib/jhipster/index.ts';
import type { WorkspacesApplication } from '../types.js';

const { PROMETHEUS, ELK } = monitoringTypes;
const { NO: NO_SERVICE_DISCOVERY, CONSUL, EUREKA } = serviceDiscoveryTypes;

/**
 * @param {Object} config - config to load config from
 * @param {import('./base-application/types.js').PlatformApplication} dest - destination context to use default is context
 */
export const loadPlatformConfig = ({ config, application }: { config: any; application: WorkspacesApplication }) => {
  mutateData(application, pickFields(config, ['serviceDiscoveryType', 'monitoring']));
};

export const loadDerivedServerAndPlatformProperties = ({ application }: { application: any }) => {
  if (!application.serviceDiscoveryType) {
    application.serviceDiscoveryType = NO_SERVICE_DISCOVERY;
  }
  application.serviceDiscoveryAny = application.serviceDiscoveryType !== NO_SERVICE_DISCOVERY;
  application.serviceDiscoveryConsul = application.serviceDiscoveryType === CONSUL;
  application.serviceDiscoveryEureka = application.serviceDiscoveryType === EUREKA;
};

/**
 * @param {import('./bootstrap-application-server/types').SpringBootApplication} dest - destination context to use default is context
 * @param {import('./base-application/types.js').PlatformApplication} dest - destination context to use default is context
 */
export const loadDerivedPlatformConfig = ({ application }: { application: WorkspacesApplication }) => {
  application.monitoringElk = application.monitoring === ELK;
  application.monitoringPrometheus = application.monitoring === PROMETHEUS;
  loadDerivedServerAndPlatformProperties({ application });
};
