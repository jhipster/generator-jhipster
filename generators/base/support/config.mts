/**
 * Copyright 2013-2023 the original author or authors from the JHipster project.
 *
 * This file is part of the JHipster project, see https://www.jhipster.tech/
 * for more information.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

const filterNullishValues = value => value !== undefined && value !== null;

/**
 * Copy and remove null and undefined values
 * @param object
 * @returns
 */
// eslint-disable-next-line import/prefer-default-export
export function removeFieldsWithNullishValues(object: Record<string, any>): Record<string, any> {
  return filterValue(object, filterNullishValues);
}
/**
 * Copy and remove null and undefined values
 * @param object
 * @returns
 */
// eslint-disable-next-line import/prefer-default-export, @typescript-eslint/no-explicit-any
function filterValue(object: Record<string, any>, filterValue: (any) => boolean = filterNullishValues): Record<string, any> {
  const clone = {};
  for (const [key, value] of Object.entries(object)) {
    if (filterValue(value)) {
      if (typeof value === 'object') {
        if (Array.isArray(value)) {
          clone[key] = value.filter(filterValue);
        } else {
          clone[key] = removeFieldsWithNullishValues(value);
        }
      } else {
        clone[key] = value;
      }
    }
  }
  return clone;
}
