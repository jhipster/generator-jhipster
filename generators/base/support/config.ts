/**
 * Copyright 2013-2023 the original author or authors from the JHipster project.
 *
 * This file is part of the JHipster project, see https://www.jhipster.tech/
 * for more information.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

const filterNullishValues = value => value !== undefined && value !== null;

/**
 * Copy and remove null and undefined values
 * @param object
 * @returns
 */
// eslint-disable-next-line import/prefer-default-export
export function removeFieldsWithNullishValues(object: Record<string, any>): Record<string, any> {
  return filterValue(object, filterNullishValues);
}
/**
 * Copy and remove null and undefined values
 * @param object
 * @returns
 */
// eslint-disable-next-line import/prefer-default-export, @typescript-eslint/no-explicit-any
function filterValue(object: Record<string, any>, filterValue: (any) => boolean = filterNullishValues): Record<string, any> {
  const clone = {};
  for (const [key, value] of Object.entries(object)) {
    if (filterValue(value)) {
      if (typeof value === 'object') {
        if (Array.isArray(value)) {
          clone[key] = value.filter(filterValue);
        } else {
          clone[key] = removeFieldsWithNullishValues(value);
        }
      } else {
        clone[key] = value;
      }
    }
  }
  return clone;
}

/**
 * Picks every field from source.
 * A field with undefined value is returned for missing fields.
 */
export const pickFields = (source: Record<string | number, any>, fields: (string | number)[]) =>
  Object.fromEntries(fields.map(field => [field, source[field]]));

/**
 * Mutation properties accepts:
 * - functions: receives the application and the return value is set at the application property.
 * - non functions: application property will receive the property in case current value is undefined.
 *
 * Applies each mutation object in order.
 *
 * @example
 * // application = { prop: 'foo-bar', prop2: 'foo2' }
 * mutateData(
 *   data,
 *   { prop: 'foo', prop2: ({ prop }) => prop + 2 },
 *   { prop: ({ prop }) => prop + '-bar', prop2: 'won\'t override' },
 * );
 */
export const mutateData = (context: Record<string | number, any>, ...mutations: Record<string | number, any>[]) => {
  for (const mutation of mutations) {
    for (const [key, value] of Object.entries(mutation)) {
      if (typeof value === 'function') {
        context[key] = value(context);
      } else if (context[key] === undefined) {
        context[key] = value;
      }
    }
  }
};
