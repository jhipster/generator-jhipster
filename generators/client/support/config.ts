import { getFrontendAppName } from '../../base/support/basename.js';
import { CLIENT_MAIN_SRC_DIR, CLIENT_TEST_SRC_DIR } from '../../generator-constants.js';

/**
 * Load client configs into application.
 */
export const loadClientConfig = ({ config, application }: { config: any; application: any }) => {
  (application as any).clientPackageManager = config.clientPackageManager;
  application.clientFramework = config.clientFramework;
  (application as any).clientTheme = config.clientTheme;
  (application as any).clientThemeVariant = config.clientThemeVariant;
  (application as any).devServerPort = config.devServerPort;

  (application as any).clientRootDir = config.clientRootDir ?? '';
  (application as any).clientSrcDir = config.clientSrcDir ?? `${application.clientRootDir}${CLIENT_MAIN_SRC_DIR}`;
  (application as any).clientTestDir = config.clientTestDir ?? `${application.clientRootDir}${CLIENT_TEST_SRC_DIR}`;
};

/**
 * Load client derived properties.
 */
export const loadDerivedClientConfig = ({ application }: { application: any }) => {
  if ((application as any).microfrontend === undefined) {
    if ((application as any).applicationTypeMicroservice) {
      (application as any).microfrontend = application.clientFrameworkAny;
    } else if ((application as any).applicationTypeGateway) {
      (application as any).microfrontend = (application as any).microfrontends && (application as any).microfrontends.length > 0;
    }
  }
  (application as any).clientThemeNone = (application as any).clientTheme === 'none';
  (application as any).clientThemePrimary = (application as any).clientThemeVariant === 'primary';
  (application as any).clientThemeLight = (application as any).clientThemeVariant === 'light';
  (application as any).clientThemeDark = (application as any).clientThemeVariant === 'dark';

  if ((application as any).baseName) {
    (application as any).frontendAppName = getFrontendAppName({ baseName: (application as any).baseName });
  }
};
