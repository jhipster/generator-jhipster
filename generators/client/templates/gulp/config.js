module.exports = {
    app: '<%= MAIN_SRC_DIR %>',
    dist: '<%= DIST_DIR %>',
    test: '<%= TEST_SRC_DIR %>'<% if(useSass) { %>,
    scss: '<%= MAIN_SRC_DIR %>scss/',
    sassSrc: '<%= MAIN_SRC_DIR %>scss/**/*.{scss,sass}',
    sassVendor: 'src/main/webapp/scss/vendor.scss',
    cssDir: '<%= MAIN_SRC_DIR %>content/css'<% } %>,
    bower: '<%= MAIN_SRC_DIR %>bower_components/',
    tmp: '<%= BUILD_DIR %>tmp',
    revManifest: '<%= BUILD_DIR %>tmp/rev-manifest.json',
    port: 9000,
    apiPort: <%= serverPort %>,
    liveReloadPort: 35729,
    uri: 'http://localhost:'
};
