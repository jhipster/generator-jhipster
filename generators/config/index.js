/**
 * Copyright 2013-2019 the original author or authors from the JHipster project.
 *
 * This file is part of the JHipster project, see https://www.jhipster.tech/
 * for more information.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
/* eslint-disable consistent-return */
const chalk = require('chalk');
const debug = require('debug')('jhipster:config');

const BaseBlueprintGenerator = require('../generator-base-blueprint');
const appPrompts = require('../app/prompts');
const clientPrompts = require('../client/prompts');
const serverPrompts = require('../server/prompts');
const { getBase64Secret, getRandomHex } = require('../utils');

// Migrate to local variables with yeoman-generator 4.1.1 | 4.2.0
module.exports = class extends BaseBlueprintGenerator {
    constructor(args, opts) {
        super(args, opts);

        debug(`Initializing ${this.rootGeneratorName()}:config generator`);

        this.generatorSource = this.options.generatorSource;
        if (!this.generatorSource) {
            this.error('Config module must have a generator source.');
        }

        if (this.generatorSource.options.namespace.endsWith(':app')) {
            this.generatorType = 'app'; // Use for languages generators only
            this.isAppConfiguration = true;
        } else if (this.generatorSource.options.namespace.endsWith(':client')) {
            this.generatorType = 'client';
            this.isClientConfiguration = true;
        } else if (this.generatorSource.options.namespace.endsWith(':server')) {
            this.generatorType = 'server';
            this.isServerConfiguration = true;
        } else {
            this.error(`Config module not implemented for ${this.generatorSource.options.namespace}.`);
        }
        debug(`Executing config for ${this.generatorType}`);

        this.serverExistingProject = this._serverExistingProject();

        this.useBlueprints = !opts.fromBlueprint && this.instantiateBlueprints('config', { generatorSource: this.generatorSource });

        this.loadScopedOptions('storage', ['base-name', 'uaa-base-name', 'skip-user-management', 'rememberMeKey']);
    }

    // Public API method used by the getter and also by Blueprints
    _initializing() {
        return {
            updateAngularName() {
                if (this.storedConfig.clientFramework === 'angular' || this.storedConfig.clientFramework === 'angular2') {
                    /* for backward compatibility */
                    this.storedConfig.clientFramework = 'angularX';
                }
            },

            validateFromCli() {
                this.checkInvocationFromCLI();
            },

            loadDBFromOptions() {
                // Override existing db by the one from cli arguments.
                if (this.options.db) {
                    const config = this.storedConfig;
                    config.databaseType = this.getDBTypeFromDBValue(this.options.db);
                    config.devDatabaseType = this.options.db;
                    config.prodDatabaseType = this.options.db;
                }
            }
        };
    }

    get initializing() {
        if (this.useBlueprints) return;
        return this._initializing();
    }

    _prompting() {
        const config = this.storedConfig;

        let clientSteps;
        if (!config.skipServer)
            clientSteps = {
                askForModuleName: this.storedConfig.baseName ? undefined : this.askModuleName,
                askForClient: clientPrompts.askForClient,
                askFori18n: this.isClientConfiguration ? this.aski18n : undefined,
                askForClientTheme: clientPrompts.askForClientTheme,
                askForClientThemeVariant: clientPrompts.askForClientThemeVariant
            };

        let serverSteps;
        if (!config.skipServer)
            serverSteps = {
                askForModuleName: this.storedConfig.baseName ? undefined : this.askModuleName,
                askForServerSideOpts: serverPrompts.askForServerSideOpts,
                configureServerPrompt: this._configureServerPrompt,
                askForOptionalItems: serverPrompts.askForOptionalItems,
                askFori18n: this.isServerConfiguration ? this.aski18n : undefined
            };

        let steps;

        // App configuration steps
        if (!this.configExisted) {
            if (this.isAppConfiguration) {
                steps = {
                    askForInsightOptIn: appPrompts.askForInsightOptIn,
                    askForApplicationType: appPrompts.askForApplicationType,
                    parseApplicationType: this._parseApplicationType,
                    askForModuleName: this.askModuleName,
                    ...serverSteps,
                    ...clientSteps,
                    askFori18n: this.skipI18n ? undefined : this.aski18n,
                    askForTestOpts: appPrompts.askForTestOpts,
                    askForMoreModules: appPrompts.askForMoreModules
                };
            } else if (this.isClientConfiguration) {
                steps = clientSteps;
            } else if (this.isServerConfiguration) {
                steps = serverSteps;
            }
        } else if (this.isAppConfiguration) {
            steps = { askForInsightOptIn: appPrompts.askForInsightOptIn };
            if (!this.serverExistingProject) {
                steps = {
                    ...steps,
                    ...serverSteps
                };
            }
        }

        return {
            ...steps,

            composeLanguages() {
                const config = this.storedConfig;
                // Update generatorType for languages generators
                let generatorType = this.generatorType;
                if (generatorType === 'app') {
                    if (config.skipClient) {
                        generatorType = 'server';
                    } else if (config.skipServer) {
                        generatorType = 'client';
                    } else {
                        this._validateSkip();
                    }
                }
                this.languages = config.languages;
                this.enableTranslation = config.enableTranslation;
                this.composeLanguagesSub(this, this.configOptions, generatorType);
            }
        };
    }

    get prompting() {
        if (this.useBlueprints) return;
        return this._prompting();
    }

    _configuring() {
        let steps;

        // App configuration steps
        if (!this.isClientConfiguration && !this.storedConfig.skipServer) {
            steps = {
                configureServer: this._configureServer
            };
        }

        return {
            ...steps,

            configuringAuthenticationType() {
                const config = this.storedConfig;
                if (config.authenticationType === 'oauth2' || (config.databaseType === 'no' && config.authenticationType !== 'uaa')) {
                    config.skipUserManagement = true;
                }
            },

            validateTranslation() {
                const config = this.storedConfig;
                config.enableI18nRTL = this.isI18nRTLSupportNecessary(config.languages);

                if (config.nativeLanguage === undefined) {
                    config.nativeLanguage = 'en';
                }
                if (config.enableTranslation === undefined) {
                    config.enableTranslation = true;
                }
                if (config.enableTranslation && config.languages === undefined) {
                    config.languages = ['en', 'fr'];
                }
            },
            validateSkip: this._validateSkip
        };
    }

    get configuring() {
        if (this.useBlueprints) return;
        return this._configuring();
    }

    /* ======================================================================== */
    /* private methods use within generator (not queued for execution) */
    /* ======================================================================== */
    _parseApplicationTypeApp() {
        const config = this.storedConfig;
        if (config.applicationType === 'microservice') {
            config.skipClient = true;
            config.skipUserManagement = true;
        }
        if (config.applicationType === 'uaa') {
            config.skipClient = true;
            config.skipUserManagement = false;
            config.authenticationType = 'uaa';
        }
    }

    _validateSkip() {
        const generator = this;
        if (generator.storedConfig.skipServer && generator.storedConfig.skipClient) {
            generator.error(`You can not pass both ${chalk.yellow('--skip-client')} and ${chalk.yellow('--skip-server')} together`);
        }
    }

    _serverExistingProject() {
        const config = this.storedConfig;
        // Keep old existingProject logic.
        const serverConfigFound =
            config.packageName !== undefined &&
            config.authenticationType !== undefined &&
            config.cacheProvider !== undefined &&
            config.enableHibernateCache !== undefined &&
            config.websocket !== undefined &&
            config.databaseType !== undefined &&
            config.devDatabaseType !== undefined &&
            config.prodDatabaseType !== undefined &&
            config.searchEngine !== undefined &&
            config.buildTool !== undefined &&
            config.baseName !== undefined;

        if (serverConfigFound) {
            this.log(
                chalk.green(
                    'This is an existing project, using the configuration from your .yo-rc.json file \nto re-generate the project...\n'
                )
            );
            return true;
        }
        return false;
    }

    _configureServerPrompt() {
        if (this.serverExistingProject) return;
        const config = this.storedConfig;

        // JWT authentication is mandatory with Eureka, so the JHipster Registry
        // can control the applications
        if (config.serviceDiscoveryType === 'eureka' && config.authenticationType !== 'uaa' && config.authenticationType !== 'oauth2') {
            config.authenticationType = 'jwt';
        }

        if (config.authenticationType === 'session') {
            config.rememberMeKey = getRandomHex();
        }

        if (config.authenticationType === 'jwt' || config.applicationType === 'microservice') {
            config.jwtSecretKey = getBase64Secret(null, 64);
        }

        // user-management will be handled by UAA app, oauth expects users to be managed in IpP
        if ((config.applicationType === 'gateway' && config.authenticationType === 'uaa') || config.authenticationType === 'oauth2') {
            config.skipUserManagement = true;
        }

        if (config.applicationType === 'uaa') {
            config.authenticationType = 'uaa';
        }

        if (config.reactive) {
            config.cacheProvider = 'no';
        }
        if (config.serverPort === undefined) {
            config.serverPort = '8080';
        }
        if (config.databaseType === 'no') {
            config.devDatabaseType = 'no';
            config.prodDatabaseType = 'no';
            config.enableHibernateCache = false;
            if (config.authenticationType !== 'uaa') {
                config.skipUserManagement = true;
            }
        } else if (config.databaseType === 'mongodb') {
            config.devDatabaseType = 'mongodb';
            config.prodDatabaseType = 'mongodb';
            config.enableHibernateCache = false;
        } else if (config.databaseType === 'couchbase') {
            config.devDatabaseType = 'couchbase';
            config.prodDatabaseType = 'couchbase';
            config.enableHibernateCache = false;
        } else if (config.databaseType === 'cassandra') {
            config.devDatabaseType = 'cassandra';
            config.prodDatabaseType = 'cassandra';
            config.enableHibernateCache = false;
        }
    }

    _configureServer() {
        const config = this.storedConfig;
        if (config.websocket === 'no') {
            config.websocket = false;
        }
        if (config.searchEngine === 'no') {
            config.searchEngine = false;
        }
        if (config.messageBroker === 'no') {
            config.messageBroker = false;
        }

        if (config.serviceDiscoveryType === 'no') {
            config.serviceDiscoveryType = false;
        }

        config.cacheProvider = config.cacheProvider || config.hibernateCache || 'no';
        config.enableHibernateCache = config.enableHibernateCache && !['no', 'memcached'].includes(config.cacheProvider);

        if (config.databaseType === 'mongodb') {
            config.devDatabaseType = 'mongodb';
            config.prodDatabaseType = 'mongodb';
            config.enableHibernateCache = false;
        } else if (config.databaseType === 'couchbase') {
            config.devDatabaseType = 'couchbase';
            config.prodDatabaseType = 'couchbase';
            config.enableHibernateCache = false;
        } else if (config.databaseType === 'cassandra') {
            config.devDatabaseType = 'cassandra';
            config.prodDatabaseType = 'cassandra';
            config.enableHibernateCache = false;
        } else if (config.databaseType === 'no') {
            config.devDatabaseType = 'no';
            config.prodDatabaseType = 'no';
            config.enableHibernateCache = false;
            if (config.authenticationType !== 'uaa') {
                config.skipUserManagement = true;
            }
        }

        // force variables unused by microservice applications
        if (config.applicationType === 'microservice' || config.applicationType === 'uaa') {
            config.websocket = false;
        }

        if (config.entitySuffix === config.dtoSuffix) {
            this.error('Entities cannot be generated as the entity suffix and DTO suffix are equals !');
        }

        // Generate remember me key if key does not already exist in config
        if (config.authenticationType === 'session' && config.rememberMeKey === undefined) {
            config.rememberMeKey = getRandomHex();
        }

        // Generate JWT secret key if key does not already exist in config
        if (config.authenticationType === 'jwt' && config.jwtSecretKey === undefined) {
            config.jwtSecretKey = getBase64Secret(null, 64);
        }

        // user-management will be handled by UAA app, oauth expects users to be managed in IpP
        if ((config.applicationType === 'gateway' && config.authenticationType === 'uaa') || config.authenticationType === 'oauth2') {
            config.skipUserManagement = true;
        }

        config.packageFolder = config.packageName.replace(/\./g, '/');

        if (config.authenticationType === 'uaa' && !config.uaaBaseName) {
            const uaaAppData = this.getUaaAppName('../uaa');
            config.uaaBaseName = uaaAppData ? uaaAppData.baseName : undefined;
        }
    }
};
