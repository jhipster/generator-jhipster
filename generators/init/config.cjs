/**
 * Copyright 2013-2022 the original author or authors from the JHipster project.
 *
 * This file is part of the JHipster project, see https://www.jhipster.tech/
 * for more information.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
const {
  PRETTIER_DEFAULT_INDENT,
  PRETTIER_DEFAULT_INDENT_DEFAULT_VALUE,
  SKIP_COMMIT_HOOK,
  SKIP_COMMIT_HOOK_DEFAULT_VALUE,
} = require('./constants.cjs');

/** Config required at .yo-rc.json */
const requiredConfig = {};

/** Default config for templates */
const defaultConfig = {
  ...requiredConfig,
  [PRETTIER_DEFAULT_INDENT]: PRETTIER_DEFAULT_INDENT_DEFAULT_VALUE,
  [SKIP_COMMIT_HOOK]: SKIP_COMMIT_HOOK_DEFAULT_VALUE,
};

module.exports = { requiredConfig, defaultConfig };
