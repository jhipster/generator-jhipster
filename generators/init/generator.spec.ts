/**
 * Copyright 2013-2025 the original author or authors from the JHipster project.
 *
 * This file is part of the JHipster project, see https://www.jhipster.tech/
 * for more information.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { basename, dirname } from 'path';
import { fileURLToPath } from 'url';
import { before, describe, expect, it } from 'esmocha';
import { basicTests, getCommandHelpOutput, testBlueprintSupport } from '../../test/support/tests.js';
import { defaultHelpers as helpers, result } from '../../lib/testing/index.js';
import { GENERATOR_INIT } from '../generator-list.js';
import { defaultConfig, requiredConfig } from './config.js';

const __filename = fileURLToPath(import.meta.url);
const __dirname = dirname(__filename);

const generator = basename(__dirname);

describe(`generator - ${generator}`, () => {
  it('generator-list constant matches folder name', () => {
    expect(GENERATOR_INIT).toBe(generator);
  });
  basicTests({
    generatorNamespace: generator,
    requiredConfig,
    defaultConfig,
    customPrompts: {},
  });
  describe('blueprint support', () => testBlueprintSupport(generator));
  describe('help', () => {
    it('should print expected information', async () => {
      expect(await getCommandHelpOutput(generator)).toMatchSnapshot();
    });
  });
  describe('with', () => {
    describe('default config', () => {
      before(async () => {
        await helpers
          .runJHipster('init')
          .withMockedJHipsterGenerators(['bootstrap'])
          .withSharedApplication({ projectDescription: 'projectDescription', prettierTabWidth: 'prettierTabWidth' })
          .withJHipsterConfig();
      });
      it('should write files and match snapshot', () => {
        expect(result.getStateSnapshot()).toMatchSnapshot();
      });

      it('should compose with generators', () => {
        expect(result.getComposedGenerators()).toMatchInlineSnapshot(`
[
  "jhipster:git",
  "jhipster:javascript:bootstrap",
  "jhipster:javascript:eslint",
  "jhipster:javascript:husky",
  "jhipster:javascript:prettier",
]
`);
      });
    });
  });
});
