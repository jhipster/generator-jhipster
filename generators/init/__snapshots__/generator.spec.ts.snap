// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generator - init help should print expected information 1`] = `
"Usage: jhipster init [options]

Init project (alpha)

Options:
  --auto-crlf                   Detect line endings
  --skip-cache                  Do not remember prompt answers (default: false)
  --skip-install                Do not automatically install dependencies (default: false)
  --force-install               Fail on install dependencies error (default: false)
  --ask-answered                Show prompts for already configured options (default: false)
  --skip-git                    Skip git repository initialization
  --force-git                   Force commit to git repository
  --commit-msg <value>          Commit changes (implies forceGit)
  --monorepository              Use monorepository
  --base-name <value>           Application base name
  --prettier-tab-width <value>  Default tab width for prettier (default: 2)
  --skip-commit-hook            Skip adding husky commit hooks
  -h, --help                    display help for command
"
`;

exports[`generator - init with default config should write files and match snapshot 1`] = `
{
  ".editorconfig": {
    "stateCleared": "modified",
  },
  ".gitattributes": {
    "stateCleared": "modified",
  },
  ".gitignore": {
    "stateCleared": "modified",
  },
  ".yo-rc.json": {
    "stateCleared": "modified",
  },
  "README.md": {
    "stateCleared": "modified",
  },
}
`;
